---
title: "Interpreting Numbat outputs"
documentclass: book
output:
  bookdown::gitbook: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

In this tutorial, we will illustrate how to visualize and interpret Numbat output, using a triple-negative breast cancer dataset (TNBC1) from [Gao et al](https://www.nature.com/articles/s41587-020-00795-2). We will use `pagoda2` for visualizing cells in low-dimensional expression space.

```{r}
devtools::load_all('~/Numbat')
library(ggplot2)
# library(Numbat)
library(dplyr)
library(conos)
library(glue)
library(data.table)
library(ggtree)
library(patchwork)
```

```{r}
# summarize output files into a Numbat object
nb = Numbat$new(out_dir = '~/paper_data/numbat_out/TNBC1_demo')
# read saved pagoda2 object
pagoda = readRDS('~/paper_data/conos_objects/TNBC1.rds')
```

## Copy number landscape and single-cell phylogeny
As an overview, we can visualize the CNV calls in single-cells and their evolutionary relationships in an integrated view:
```{r, fig.width = 8, fig.height=3.5, dpi=300}
mypal = c('gray', "#377EB8", "#4DAF4A", "#984EA3")
  
nb$plot_phylo_heatmap(
  geno_bar = TRUE, 
  p_min = 0.9,
  pal_clone = mypal,
  multi_allelic = TRUE
)
```
In this visualization, the single-cell phylogeny (left) is juxtaposed with a heatmap of single-cell CNV calls (right). The CNV calls are colored by the type of alterations (AMP, amplification, BAMP, balanced amplification, DEL, deletion, CNLoH, copy-neutral loss of heterozygosity). The colorbar in-between differentiates the distinct genetic populations (genotype). This tells us that the dataset mainly consists of three cell populations, a normal population (gray) and two tumor subclones (green and yellow). 

## Single-cell CNV calls
Numbat probabilistically evaluates the presence/absence of CNVs in single cells. The cell-level CNV posteriors are stored in the `nb$joint_post` dataframe:
```{r}
head(nb$joint_post) %>% select(cell, CHROM, seg, cnv_state, p_cnv, p_cnv_x, p_cnv_y)
```
which contains cell-level information on specific CNV segments (`seg`), their alteration type (`cnv_state`), the joint posterior probability of the CNV (`p_cnv`), the expression-based posterior (`p_cnv_x`), and the allele-based posterior (`p_cnv_y`). We can visualize the event-specific posteriors in a expression-based tSNE embedding:
```{r, fig.width = 9, fig.height=3, dpi=300}

plist = list()

muts = c('1b', '3c', '4c')
# muts = unique(nb$joint_post$seg_label) %>% gtools::mixedsort()
cnv_type = nb$joint_post %>% distinct(seg, cnv_state) %>% {setNames(.$cnv_state, .$seg)}

for (mut in muts) {
    
    plist[[mut]] = pagoda$plotEmbedding(
        alpha=0.8,
        size=1, 
        plot.na = F, 
        colors = nb$joint_post %>%
            filter(seg == mut) %>%
            {setNames(.$p_cnv, .$cell)},
        show.legend = T,
        mark.groups = F,
        plot.theme = theme_bw(),
        title = paste0(mut, '(', cnv_type, ')')
    ) +
    scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, limits = c(0,1))
}

wrap_plots(plist, guides = 'collect')
```

## Tumor phylogeny
We can get a more detailed view of the reconstructed single-cell phylogeny as below.
```{r, fig.width = 5, fig.height = 3, dpi=300}

mut_nodes = nb$gtree %>% activate(nodes) %>%
    filter(!is.na(site)) %>%
    data.frame() %>%
    select(name, site)

branch_width = 1

nb$gtree %>% 
  to_phylo() %>%
  ggtree(ladderize = T, size = branch_width) %<+%
  mut_nodes +
  layout_dendrogram() +
  theme(
      plot.margin = margin(0,0,0,0),
      axis.ticks.x = element_blank(),
      axis.text.x = element_blank(),
      axis.line.y = element_line(size = 0.2),
      axis.ticks.y = element_line(size = 0.2),
      axis.text.y = element_text(),
      axis.title.x = element_text()
  ) +
  guides(color = F)  +
  geom_point2(aes(subset = !is.na(site), x = branch), shape = 21, size = 1, fill = 'red') +
  geom_text2(
      aes(x = branch, label = str_trunc(site, 20, side = 'center')),
      size = 2, hjust = 0, vjust = -0.5, nudge_y = 1, color = 'darkred'
  ) +
  xlab('Number of mutations')

```

The clone-level mutational history constructed by Numbat can be visualized. Here the nodes represent clones (each carrying a distinct genotype) and edges represent mutational events. 
```{r, fig.width = 6, fig.height=3, dpi=300}
nb$plot_mut_history(pal_clone = mypal)
```

## Clonal assignments
Numbat aggregates signals across subclone-specific CNVs to probabilistically assign cells to subclones. The information regarding clonal assignments are contained in the `nb$clone_post` dataframe.
```{r}
nb$clone_post %>% head() %>% select(cell, clone_opt, p_1, p_2, p_3, p_4)
```
Here `clone_opt` denotes the maximum likelihood assignment of a cell to a specific clone. `p_{1..4}` are the detailed breakdown of the posterior probability that the cell belongs to each clone, respectively. Let's visualize the clonal decomposition in a tSNE embedding.
```{r, fig.width = 3, fig.height=3, dpi=300}
pagoda$plotEmbedding(
    alpha=0.8,
    size=1, 
    groups = nb$clone_post %>%
        {setNames(.$clone_opt, .$cell)},
    plot.na = F,
    plot.theme = theme_bw(),
    title = 'Genotypes',
    pal = mypal
)
```

## Tumor versus normal probability
Combining evidence from all CNVs, Numbat derives an aneuploidy probability for each cell to distinguish tumor versus normal cells. We can visualize the posterior aneuploidy probability based on expression evidence only, allele evidence only, and jointly:
```{r, fig.width = 9, fig.height=3, dpi=300}

p_joint = pagoda$plotEmbedding(
    alpha=0.8,
    size=1, 
    colors = nb$clone_post %>%
        {setNames(.$p_cnv, .$cell)},
    plot.na = F,
    plot.theme = theme_bw(),
    title = 'Joint',
) +
scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5)

p_allele = pagoda$plotEmbedding(
    alpha=0.8,
    size=1, 
    colors = nb$clone_post %>%
        {setNames(.$p_cnv_x, .$cell)},
    plot.na = F,
    plot.theme = theme_bw(),
    title = 'Expnbsion',
) +
scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5)

p_expr = pagoda$plotEmbedding(
    alpha=0.8,
    size=1, 
    colors = nb$clone_post %>%
        {setNames(.$p_cnv_y, .$cell)},
    plot.na = F,
    show.legend = T,
    plot.theme = theme_bw(),
    title = 'Allele',
) +
scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5)

(p_expr | p_allele | p_joint) + plot_layout(guides = 'collect')

```
Both expression and allele signal clearly separate the tumor and normal cells.

## Bulk CNV profiles
To get a better sense of the raw CNV signals, we can visualize the pseudobulk CNV profiles of the respective clones by the below function:
```{r, fig.width = 14, fig.height=6, dpi=300}
nb$plot_bulks()
```


